### PR TITLE
Add error handling

### PR DESCRIPTION
**Pull request by evanw555**
Preliminary review, please take a look and let me know what you think.

### CODE REVIEW

File: restlicsharpclient/restliclient/transport/RestliCallbackAdapter.cs [hunk 1]
```diff
@@ -39,8 +39,17 @@ public RestliCallbackAdapter(RestResponseDecoder<TResponse> responseDecoder, Res
...
+                callback.OnError(clientErrorResponse);
+            }
+            else
+            {
+                TResponse response = responseDecoder.DecodeResponse(transportResponse);
+                callback.OnSuccess(response);
+            }
         }
 
         public void OnError(HttpResponse httpResponse)
```
Comment (mchen07) [inline restlicsharpclient/restliclient/transport/RestliCallbackAdapter.cs]: This seems not used at all. I am suggesting that we can change TransportCallback interface only have one method: onResponse(HttpResponse).

File: restlicsharpclient/restliclient/util/DataUtil.cs [hunk 2]
```diff
@@ -130,7 +130,16 @@ public static byte[] MapToBytes(Dictionary<string, object> dataMap)
...
         public static Dictionary<string, object> BytesToMap(byte[] dataBytes)
         {
-            return StringToMap(System.Text.Encoding.UTF8.GetString(dataBytes));
+            if (dataBytes == null)
+            {
+                return null;
+            }
+            else
+            {
+                string stringData = System.Text.Encoding.UTF8.GetString(dataBytes);
```
Comment (evanw555) [inline restlicsharpclient/restliclient/util/DataUtil.cs]: Oops, I was doing some debugging and forgot to remove that.

File: restlicsharpclient/restliclientintegrationtest/RestClientIntegrationTests.cs [hunk 3]
```diff
@@ -98,14 +100,45 @@ public void GetGreeting_Async()
...
+            };
+            RestliCallback<EntityResponse<Greeting>> callback = new RestliCallback<EntityResponse<Greeting>>(successHandler, errorHandler);
+
+            client.RestRequestAsync(request, callback);
+
+            blocker.WaitOne();
+
+            Assert.IsNull(greeting);
+            Assert.IsNotNull(error);
+            Assert.IsNotNull(error.details);
```
Comment (mchen07) [inline restlicsharpclient/restliclientintegrationtest/RestClientIntegrationTests.cs]: Please also verify status code and message in error.details are value returned from server.

File: restlicsharpclient/restliclientintegrationtest/RestClientIntegrationTests.cs [hunk 4]
```diff
@@ -98,14 +100,45 @@ public void GetGreeting_Async()
...
+
+            RestliCallback<EntityResponse<Greeting>>.SuccessHandler successHandler = delegate (EntityResponse<Greeting> response)
+            {
+                greeting = response.element;
+                blocker.Set();
+            };
+            RestliCallback<EntityResponse<Greeting>>.ErrorHandler errorHandler = delegate (ClientErrorResponse response)
+            {
+                error = response.error;
+                blocker.Set();
```
Comment (mchen07) [inline restlicsharpclient/restliclientintegrationtest/RestClientIntegrationTests.cs]: Also verify ClientErrorResponse httpStatus is correct.

**Pull request merged**